Algo/src/roots.cxx

```diff
@@ -115,21 +105,45 @@ extern "C" void Test_roots(CCTK_ARGUMENTS) {
   }
 
   {
-    const int minbits =
-        static_cast<int>(0.6 * std::numeric_limits<double>::digits);
-    const int maxiters = 100;
-    int iters;
-    bool failed;
-    auto x = newton_raphson_nd(gnd<CCTK_REAL>, vec<CCTK_REAL, 2>{1.0, 1.0},
-                               vec<CCTK_REAL, 2>{0.0, 0.0},
-                               vec<CCTK_REAL, 2>{10.0, 10.0}, minbits, maxiters,
-                               iters, failed);
-    // CCTK_VINFO("maxiters=%d iters=%d", maxiters, iters);
-    // CCTK_VINFO("lo=%.17g hi=%.17g", double(lo), double(hi));
-    assert(iters < maxiters);
-    assert(sumabs(gn(x)) < 1.0e-9);
-    CCTK_VINFO("Test_newton_raphson_nd succeeded in %d iterations", iters);
+    const amrex::Box box(amrex::IntVect(0), amrex::IntVect(0, 0, 0));
+
+    amrex::ParallelFor(
+        box, [=] CCTK_DEVICE(const int i, const int j, const int k)
+        CCTK_ATTRIBUTE_ALWAYS_INLINE {
+      const int minbits = std::numeric_limits<double>::digits - 4;
+      const int maxiters = 100;
+      int iters;
+      auto [lo, hi] = brent(fn, 1.0, 2.0, minbits, maxiters, iters);
+      // CCTK_VINFO("maxiters=%d iters=%d", maxiters, iters);
+      // CCTK_VINFO("lo=%.17g hi=%.17g", double(lo), double(hi));
+      assert(iters < maxiters);
+      assert(hi >= lo && hi - lo <= std::scalbn(2.0, -minbits));
+      assert(fn(lo) <= 0 && fn(hi) >= 0);
+      //CCTK_VINFO("Test_brent succeeded in %d iterations", iters);
+      printf("INFO (Algo): Test_brent succeeded in %d iterations\n", iters);
```

Comment:
Don't call `printf` from inside the kernel. Instead, check whether the kernel finished successfully outside the kernel, and then call `CCTK_INFO`.